### ISSUE TITLE
Cannot compile without GL.h

### ISSUE THREAD

**Issue by flyx** [go-gl/glfw#242]
All versions of this wrapper cannot be compiled without `GL.h`. However, this header file does not necessarily exist on systems implementing OpenGL ES.

GLFW provides [header macros](https://www.glfw.org/docs/latest/build.html#build_macros) that should be defined for such cases (e.g. `GLFW_INCLUDE_ES1`). It should be possible to configure this wrapper accordingly. Possible solutions:

### Build Tags

I tried this in `build.go`:

```c
// OpenGL API Version Build Tags
// ----------------
#cgo opengles1 CFLAGS: -DGLFW_INCLUDE_ES1
#cgo opengles2 CFLAGS: -DGLFW_INCLUDE_ES2
#cgo opengles3 CFLAGS: -DGLFW_INCLUDE_ES3
#cgo opengles31 CFLAGS: -DGLFW_INCLUDE_ES31
#cgo openglcorearb CFLAGS: -DGLFW_INCLUDE_GLCOREARB
```

However, this will lead to GLFW itself being compiled with these macros, which it doesn't like that much:

```
In file included from ./c_glfw.go:4:
In file included from ./glfw/src/context.c:28:
./glfw/src/internal.h:45:3: error: "You must not define any header option macros when compiling GLFW"
```

This is because [cgo defines](https://golang.org/cmd/cgo/#hdr-Using_cgo_with_the_go_command):

> All the cgo CPPFLAGS and CFLAGS directives in a package are concatenated and used to compile C files in that package.

So to make this work, the build of the GLFW library must be moved to another package.

### GLFW_INCLUDE_NONE

The second alternative is to manually define this every time `glfw3.h` is included:

    //#define GLFW_INCLUDE_NONE
    //#include "glfw/include/GLFW/glfw3.h"

Problem with this is that `window.go` and `util.go` reference `C.GL_TRUE` which is only defined if *some* OpenGL implementation is included. However, I think that this reference should vanish anyway because GLFW supports Vulkan and that also does not have `GL_TRUE` afaik. I think it can be safely define as a `const` on the Go side.

---

I would happily create a PR if I know which option is preferred by y'all. Note that I am new to Go and may have overlooked something obvious.

**Comment by tapir** [comment]
It's interesting that GLFW does not want to be compiled when another GL header is specified. Maybe we should ask @elmindreda why is it like that in the first place.

BTW may I ask why your system doesn't have GL.h?

**Comment by elmindreda** [comment]
Cool, wasn't sure that error message would ever be useful.

> It's interesting that GLFW does not want to be compiled when another GL header is specified. Maybe we should ask @elmindreda why is it like that in the first place.

GLFW needs to use symbols from many versions of GL, GLES and related APIs, many headers out there lack symbols it need, some of the headers conflict or aren't easily available, so [we have our own](https://github.com/glfw/glfw/blob/5644b66bac4a8e5e4f42267f68a1448a04c5e6ea/src/internal.h#L86) with only what we need. Essentially GLFW contains a very tiny custom extension loader.

**Comment by flyx** [comment]
@tapir I am developing for a [Rock64 board](https://www.pine64.org/?page_id=7147) whose [Mali GPU][1] only supports OpenGL ES. I mean, obviously I could get some `GL.h` and place it somewhere, but that would be rather hacky.

 [1]: https://en.wikipedia.org/wiki/Mali_(GPU)

**Comment by tapir** [comment]
I think `//#define GLFW_INCLUDE_NONE` is the correct solution here. For `GL_TRUE` and `GL_FALSE` we can define them as type `C.int` with values 0 and 1. If you define them as consts in Go side then you need to convert them before passing it to C side GLFW funcs thus requires modification on some wrappers.

Could you send a pull request?

**Comment by tapir** [comment]
> Cool, wasn't sure that error message would ever be useful.

Ineed. Kudos :+1:

**Comment by flyx** [comment]
@tapir I have a working fork. I figured out that I do need to have additional build tags anyway, to configure linker switches for GLFW (`-lGL` vs `-lGLESv*x*`); I also found that there are `True` and `False` consts defined already, so I used those along with conversions where necessary (I expect them to produce zero instructions when compiling).

I'll prepare a PR on the weekend. Need to ensure other environments than mine (GLESv2) will work as well.
